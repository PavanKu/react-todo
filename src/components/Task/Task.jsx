import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";

import StyledTask, {
  StyledCheckBox,
  StyledTaskLabel,
  StyledTaskDelete,
} from "./Task.styled";

export const DEFAULT_TASK_STATUS = false;
export const DEFAULT_TASK = "";
export const DEFAULT_SHOW_CLOSE = null;

function Task({
  isCompleted = DEFAULT_TASK_STATUS,
  showClose = DEFAULT_SHOW_CLOSE,
  task = DEFAULT_TASK,
  taskId,
  handleDelete,
  handleToggle,
}) {
  const [inputChecked, setInputChecked] = useState(isCompleted);
  const [isCloseVisible, setIsCloseVisible] = useState(showClose);

  useEffect(() => {
    setInputChecked(isCompleted);
  }, [isCompleted]);

  useEffect(() => {
    setIsCloseVisible(showClose);
  }, [showClose]);

  const handleInputChange = (event) => {
    if (handleToggle) {
      const isChecked = event.target.checked;
      setInputChecked(isChecked);
      handleToggle(taskId, isChecked);
    }
  };

  const handleRemove = () => {
    if (handleDelete) {
      handleDelete(taskId);
    }
  };

  const showCloseBtn = () => {
    if (showClose === null && !isCloseVisible) {
      setIsCloseVisible(true);
    }
  };
  const hideCloseBtn = () => {
    if (showClose === null && isCloseVisible) {
      setIsCloseVisible(false);
    }
  };

  return (
    <StyledTask onMouseEnter={showCloseBtn} onMouseLeave={hideCloseBtn}>
      <StyledCheckBox
        id={taskId}
        onChange={handleInputChange}
        isCompleted={inputChecked}
      />
      <StyledTaskLabel isCompleted={inputChecked} htmlFor={taskId}>
        {task}
      </StyledTaskLabel>
      {isCloseVisible && <StyledTaskDelete onClick={handleRemove} />}
    </StyledTask>
  );
}

Task.propTypes = {
  isCompleted: PropTypes.bool,
  task: PropTypes.string,
  taskId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  handleDelete: PropTypes.func,
  handleToggle: PropTypes.func,
  showClose: PropTypes.bool,
};

export default Task;
