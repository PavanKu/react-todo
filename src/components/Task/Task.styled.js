import styled, { css } from "styled-components";

export const StyledCheckBox = styled.input.attrs((props) => ({
  type: "checkbox",
  checked: props.isCompleted,
}))`
  position: relative;
  width: 0px;
  height: 0px;
  :before {
    content: "";
    position: absolute;
    width: 30px;
    height: 30px;
    border: 1px solid #f0f0f0;
    border-radius: 50%;
    top: -25px;
    left: 10px;
    cursor: pointer;
  }
  :checked:after {
    content: "";
    position: absolute;
    width: 8px;
    height: 16px;
    border-bottom: 2px solid #6cc9b7;
    border-right: 2px solid #6cc9b7;
    left: 20px;
    top: -20px;
    transform: rotate(45deg);
    cursor: pointer;
  }
`;
export const StyledTaskLabel = styled.label`
  word-break: break-all;
  padding: 15px 15px 15px 60px;
  display: inline-block;
  line-height: 1.2;
  transition: color 0.4s;
  cursor: pointer;
  ${(props) =>
    props.isCompleted &&
    css`
      color: #d9d9d9;
      text-decoration: line-through;
    `}
`;
export const StyledTaskDelete = styled.button`
  position: absolute;
  top: 0;
  right: 10px;
  bottom: 0;
  width: 40px;
  height: 40px;
  margin: auto 0;
  font-size: 30px;
  color: #cc9a9a;
  margin-bottom: 11px;
  transition: color 0.2s ease-out;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
  :after {
    content: "Ã—";
  }
`;

const StyledTask = styled.div`
  position: relative;
  font-size: 24px;
  border-bottom: 1px solid #ededed;
  background: #fefefe;
`;

export default StyledTask;
