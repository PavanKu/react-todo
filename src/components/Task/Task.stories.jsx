import React from "react";
import { storiesOf } from "@storybook/react";
import { withKnobs, text, boolean } from "@storybook/addon-knobs";
import { action } from "@storybook/addon-actions";
import Task from "./Task";

const SAMPLE_TASK = "Buy Milk";

const IncompleteTask = () => (
  <Task task={SAMPLE_TASK} taskId="0" showClose={false} />
);
const CompletedTask = () => (
  <Task task={SAMPLE_TASK} taskId="0" isCompleted={true} showClose={false} />
);

const OnHoveTask = () => (
  <Task task={SAMPLE_TASK} taskId="0" isCompleted={true} />
);
const HoveredTask = () => (
  <Task task={SAMPLE_TASK} taskId="0" isCompleted={true} showClose />
);
const Playground = () => (
  <Task
    task={text("task", SAMPLE_TASK)}
    taskId="1"
    isCompleted={boolean("isCompleted", true)}
    showClose={boolean("showClose", true)}
    handleDelete={action("Task:handleDelete:")}
    handleToggle={action("Task:handleToggle:")}
  />
);
storiesOf("Task", module)
  .addDecorator(withKnobs)
  .add("Incomplete Task", IncompleteTask)
  .add("Completed Task", CompletedTask)
  .add("On Hovered Task", OnHoveTask)
  .add("Hovered Task", HoveredTask)
  .add("Playground", Playground);
