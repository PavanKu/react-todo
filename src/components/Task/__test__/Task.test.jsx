import React from "react";
import { cleanup, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Task, { DEFAULT_TASK, DEFAULT_TASK_STATUS } from "../Task";
import {
  ARIA_ROLE_CHECKBOX,
  ARIA_ROLE_BUTTON,
} from "../../../constants/ariaConstants";

describe("<Task />", () => {
  afterEach(cleanup);

  it("should render without crash", () => {
    render(<Task taskId="1" />);
  });

  it("should render with default props", () => {
    const { getByRole, getByLabelText } = render(<Task taskId="1" />);

    const checkboxElement = getByRole(ARIA_ROLE_CHECKBOX);
    const labelElement = getByLabelText(DEFAULT_TASK);

    expect(labelElement).toBeDefined();
    expect(checkboxElement.checked).toEqual(DEFAULT_TASK_STATUS);
  });

  it("should render with props", () => {
    const sampleTask = "Buy Milk";
    const { getByRole, getByLabelText } = render(
      <Task taskId="1" isCompleted showClose task={sampleTask} />
    );

    const checkboxElement = getByRole(ARIA_ROLE_CHECKBOX);
    const labelElement = getByLabelText(sampleTask);
    const closeBtnElement = getByRole(ARIA_ROLE_BUTTON);

    expect(labelElement).toBeDefined();
    expect(checkboxElement.checked).toEqual(true);
    expect(closeBtnElement).toBeDefined();
  });

  it("should invoke toggle handlers", () => {
    let taskIdFromToggle, taskStatusFromToggle;
    const sampleTask = "Buy Milk";
    const selectionChange = (id, status) => {
      taskIdFromToggle = id;
      taskStatusFromToggle = status;
    };
    const { getByRole, getByLabelText } = render(
      <Task
        taskId="1"
        showClose
        task={sampleTask}
        handleToggle={selectionChange}
      />
    );

    const checkboxElement = getByRole(ARIA_ROLE_CHECKBOX);
    const labelElement = getByLabelText(sampleTask);

    userEvent.click(checkboxElement);
    expect(taskIdFromToggle).toEqual("1");
    expect(taskStatusFromToggle).toEqual(true);

    userEvent.click(labelElement);
    expect(taskIdFromToggle).toEqual("1");
    expect(taskStatusFromToggle).toEqual(false);
  });

  it("should invoke delete handler", () => {
    let taskIdFromDelete;
    const sampleTask = "Buy Milk";

    const deleteTask = (id) => (taskIdFromDelete = id);
    const { getByRole } = render(
      <Task taskId="1" showClose task={sampleTask} handleDelete={deleteTask} />
    );

    const closeBtnElement = getByRole(ARIA_ROLE_BUTTON);

    userEvent.click(closeBtnElement);
    expect(taskIdFromDelete).toEqual("1");
  });

  it("should match snapshot", () => {
    const sampleTask = "Buy Milk";
    const { container } = render(
      <Task taskId="1" isCompleted showClose task={sampleTask} />
    );

    expect(container).toMatchSnapshot();
  });
});
