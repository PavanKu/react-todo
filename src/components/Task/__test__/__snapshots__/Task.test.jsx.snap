// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Task /> should match snapshot 1`] = `
<div>
  <div
    class="sc-AxhCb biRedW"
  >
    <input
      checked=""
      class="sc-AxjAm bLKDEa"
      id="1"
      type="checkbox"
    />
    <label
      class="sc-AxirZ ctbNrS"
      for="1"
    >
      Buy Milk
    </label>
    <button
      class="sc-AxiKw bJbxvS"
    />
  </div>
</div>
`;
